### ISSUE TITLE
Benchmark Keras Processing Layer

### ISSUE THREAD

**Issue by innat** [keras-team/keras#23245]
## System information

- Backend: TensorFlow
- GPU: NVIDIA Tesla T4 (16 GB VRAM)
- CPU RAM: 30 GB
- Batch size: 64
- Image resolution: 224 × 224
- Training mode: training=True
- Iterations: 100 (30 warmup)

## Reproducible Code

- [collab.](https://colab.research.google.com/drive/1sl5BApY9pdk_z6sDBTB3WB8dRuC9wjWn?usp=sharing)

**Disclaimer**: The code generated by claude code.

## Description

While benchmarking the built-in Keras preprocessing/augmentation layers, I noticed that several layers are unexpectedly slow on GPU and consume a large amount of VRAM compared to other augmentations.

Some layers are even slower on GPU than CPU, while others allocate hundreds of megabytes of GPU memory for relatively simple operations.

This becomes much more concerning because these benchmarks are performed under relatively small training settings:

```
batch size = 64
image size = 224×224
```

In many practice (i.e. Kaggle Competition),  high-resolution vision tasks are commonly train with resolutions such as:

```
512, 768, 1024, 2048×2048
```

The current memory footprint suggests that some augmentations may become impractical at those resolutions.

## Benchmark results

Spatial augmentations

Layer | CPU (ms) | GPU (ms) | GPU VRAM
-- | -- | -- | --
RandomFlip | 128 | 49.8 | 256 MB
RandomRotation | 129 | 56.3 | 165 MB
RandomZoom | 123 | 55.9 | 165 MB
RandomTranslation | 122 | 55.2 | 165 MB
RandomShear | 127 | 58.3 | 165 MB
RandomElasticTransform | 3328 | **3765** | 293 MB
RandomPerspective | 149 | 71.8 | 219 MB

The most notable outlier is `RandomElasticTransform`, which is extremely slow on both CPU and GPU. Surprisingly, GPU execution is even slower than CPU.

## Intensity augmentations

Layer | CPU (ms) | GPU (ms) | GPU VRAM
-- | -- | -- | --
RandomBrightness | 58 | 43 | 256 MB
RandomContrast | 110 | 44 | 256 MB
RandomColorJitter | 13.7 | 7.3 | 64 MB
RandomHue | 280 | 55 | 330 MB
RandomSaturation | 135 | 48 | 263 MB
RandomGaussianBlur | 256 | 58 | 293 MB
RandomGrayscale | 59 | 52 | 165 MB
Solarization | 108 | 52 | 293 MB

## Histogram augmentations

Layer | CPU (ms)
-- | --
Equalization | 251
CLAHE | 4671
RandomErasing | 251

Layer | GPU (ms) | GPU VRAM
-- | -- | --
Equalization | 139 | 354 MB
RandomErasing | 65 | 284 MB

`CLAHE` could not be benchmarked on GPU because it quickly led to out-of-memory errors.

## Advanced augmentations

Layer | CPU (ms) | GPU (ms) | GPU VRAM
-- | -- | -- | --
RandAugment | 3192 | 359 | 403 MB

## Mixing augmentations

Layer | CPU (ms) | GPU (ms) | GPU VRAM
-- | -- | -- | --
AugMix | 1037 | 105 | 512 MB
CutMix | 80 | 45 | 247 MB
MixUp | 54 | 32 | 293 MB

`AugMix` is particularly memory-intensive, reaching approximately 512 MB of peak GPU memory.

## Questions

1. Are these performance characteristics expected for the current implementations?
2. Are there known optimization plans for these augmentation layers?
3. Is `RandomElasticTransform` or `CLAHE` expected to be slower on GPU than CPU, or could this indicate an optimization opportunity?
4. Would it be possible to reduce temporary tensor allocations or fuse operations to improve GPU memory usage?
5. These benchmarks were performed using a standard **image classification** workload. Have the preprocessing layers also been evaluated for **object detection** and **semantic segmentation** workloads, where higher resolutions and larger feature maps are common? If so, are there any benchmark results or recommended best practices available?
6. Is there any official benchmark report available in keras.io?

Unlike traditional preprocessing pipelines, Keras preprocessing layers are not restricted to running in the data loader (CPU). Since these layers can be part of the model, they can also be executed on the GPU during training. It would therefore be very helpful if the documentation explicitly indicated which preprocessing or augmentation layers:

- are fully GPU-compatible,
- currently fall back to CPU,
- cannot yet be offloaded to GPU due to implementation or backend limitations, or
- are planned to support GPU execution in the future.

Having this information available would help users design more efficient input pipelines and make informed decisions when choosing augmentations, especially for high-resolution training workloads where GPU execution can have a significant impact on throughput.

**P.S.** Some of the questions involve implementation trade-offs, backend-specific behavior, and potential optimization opportunities, so input from a maintainer or contributor with expertise in this area would be especially valuable.

**Comment by innat** [comment]
I've just tried pygrain dataloader and keras model with its standard training API. Seems okay for all backends. However, as soon as I added `keras.layers. RandomFlip, RandomZoom, RandomRotate`, except `torch` backends, all other backends seems give similar training speed. Note, similar training speed doesn't mean 'okay', that's look a different story. I can give a reproducible code if it is needed.

**Comment by hertschuh** [comment]
@innat 

Thanks a lot for the detailed benchmarking and investigation. To answer your questions:

> 1. Are these performance characteristics expected for the current implementations?

I cannot say that any of this was expected or not expected. I don't think anybody had done the benchmarking you did.

> 2. Are there known optimization plans for these augmentation layers?

Not as of now, but it might be a good item to add to the plan.

> 3. Is RandomElasticTransform or CLAHE expected to be slower on GPU than CPU, or could this indicate an optimization opportunity?

These layers were assumed to run on CPU as part of a preprocessing pipeline like `tf.data`, so I don't think anybody thought about the performance on GPU. I'm not even sure if they actually execute on GPU vs. fallback to run on CPU.

> 4. Would it be possible to reduce temporary tensor allocations or fuse operations to improve GPU memory usage?

Since this is on TensorFlow, the first thing I would try is to compile them using a `tf.function(jit_compile=True)`. Hopefully the compiler can optimize some.

> 5. These benchmarks were performed using a standard image classification workload. Have the preprocessing layers also been evaluated for object detection and semantic segmentation workloads, where higher resolutions and larger feature maps are common? If so, are there any benchmark results or recommended best practices available?

No.

> 6. Is there any official benchmark report available in keras.io?

https://keras.io/getting_started/benchmarks/

> Unlike traditional preprocessing pipelines, Keras preprocessing layers are not restricted to running in the data loader (CPU). Since these layers can be part of the model, they can also be executed on the GPU during training. It would therefore be very helpful if the documentation explicitly indicated which preprocessing or augmentation layers:
> 
> - are fully GPU-compatible,
> - currently fall back to CPU,
> - cannot yet be offloaded to GPU due to implementation or backend limitations, or
> - are planned to support GPU execution in the future.

Agreed. But because they are implemented using Keras ops, the answer should be the same for all layers I would hope.

**Comment by innat** [comment]
@hertschuh 
Thanks for the response. 

Data augmentation on the GPU becomes especially useful under certain conditions, such as moderate input sizes, larger batch sizes, and models that do not fully utilize the available GPU resources. In these cases, GPU-based augmentation can significantly improve the overall training throughput.

This is why many practitioners have adopted this approach, it offers both performance gains and a more streamlined training pipeline. Examples include [NVIDIA DALI](https://github.com/NVIDIA/DALI), [Kornia](https://github.com/kornia/kornia), [OpenCV CUDA](https://opencv.org/platforms/cuda/) [CV-CUDA](https://github.com/CVCUDA/CV-CUDA).

In Keras, it is possible to perform GPU-based data augmentation for a variety of tasks, including classification, segmentation, and detection. For segmentation and detection in particular, users can typically enable GPU augmentation with little to no code changes while continuing to use the high-level training API (`model.fit`).

I recently ran some additional tests using the Grain dataloader across different backends and noticed that the built-in Keras augmentation layers take noticeably longer on the PyTorch backend than on the others. I think it would be helpful to keep performance in mind for similar layers or components during future PRs and reviews. Otherwise, even with a rich collection of layers, performance issues may limit their practical use during training.

I think it would be beneficial to have a dedicated issue, pinned by the maintainers, for the continuous optimization of existing components such as preprocessing and augmentation layers. There is often room to improve performance by adopting more efficient algorithms or implementations, even when the functionality remains the same.

For reference, the Colab gist shared above was only tested with the TensorFlow backend.
